db/migrate/20230605133804_remove_master_variants.rb

```diff
@@ -4,13 +4,24 @@ def change
       delete_master_option_values
     end
 
+    delete_master_line_items
     delete_master_inventory_units
     delete_master_variant_prices
     delete_master_variants
   end
 
   private
 
+  def delete_master_line_items
+    ActiveRecord::Base.connection.execute(<<-SQL
+      DELETE FROM spree_line_items
+      USING spree_variants
+      WHERE spree_variants.is_master = true
+        AND spree_variants.id = spree_line_items.variant_id
+    SQL
+    )
+  end
```

Comment:
I agree that these shouldn't be there but deleting line items sounds dangerous. We should check the impact:

```
$ sql='SELECT count(*) FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id;'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
 count 
-------
 11027
(1 row)
openfoodnetwork.org.nz | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.net | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.ca | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.org.uk | CHANGED | rc=0 >>
 count 
-------
  2229
(1 row)
app.katuma.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
coopcircuits.fr | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.ie | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
csicsoka.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.de | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.be | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
```

Only the two oldest instances have these line items. But there are many.

Reply:
Is this data ancient or do we need to worry about this?
```
$ sql='SELECT max(spree_line_items.updated_at) FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id;'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
            max             
----------------------------
 2015-06-11 08:01:57.836125
(1 row)


openfoodnetwork.org.uk | CHANGED | rc=0 >>
            max             
----------------------------
 2020-08-04 01:55:15.996018
```

2015 sounds about right but UK has such line items updated in 2020?

Reply:
I guess, if we want to play it safe, we could create a new pull request to fix this first. It could identify all these line items and point them to the first variant of the product instead of the master variant. That's maybe not 100% correct but it would keep the line items with quantity and price so that the order total makes sense.

What do you think?

Reply:
I was curious _how many_ orders there were in 2020, so I modified your query (thanks for sharing). It turns out there's just one order (at least, there's only line items created at one instant).

```
$ sql='SELECT DISTINCT spree_line_items.updated_at FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id ORDER BY spree_line_items.updated_at DESC LIMIT 20;'
$ ansible uk-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.uk | CHANGED | rc=0 >>
         updated_at
----------------------------
 2020-08-04 01:55:15.996018
 2016-03-15 20:05:51.634713
 2015-06-10 18:24:50.69657
...
```

Reply:
There's also one table that refers to spree_line_items: `spree_option_values_line_items`, so we could potentially be orphaning those. I didn't bother querying for them though, because it probably won't matter for ancient records.

Reply:
I'm not sure what to suggest. Maikel's idea of migrating makes sense and may well be perfectly appropriate, but it's hard to know without reviewing all the orders. What if the price of the first variant is different to the master variant? So at the risk of wrong data, why not just delete.

I don't think we need to keep records that old, but deleting some and not others leaves an incorrect picture of history. If someone wants to analyse past data, they'll not be getting an accurate picture. So part of me wants to suggest we communicate that we're going to delete ALL orders before those dates (for UK and AU where it's relevant).

Hmm.. dunno.. Maikel's suggestion is probably fine, and nobody will notice anyway.

Reply:
> What if the price of the first variant is different to the master variant?

The line item itself has a price stored in case you update the price of a variant later. So at least the order would be reflected correctly. There may be some information pull from the variant that could differ but the most important information would still be there through the product and the first variant.

Reply:
🤦 Ha, good point. So then it's a low risk update, and leaves data in a relatively consistent state. Seems worth doing to me.

Reply:
Hmmm it's tricky. This issue of line items that reference a master variant is a scenario that in theory shouldn't have been valid or even possible for at least 5-6 years, and these old bits of data in theory should have been cleaned up at the time those changes were made. But a couple of servers have some really old data that's not in a nice state.

I think if we go down this route of trying to switch out each of the variants in these line items to a different variant we might end up with a huge can of worms to deal with. I imagine that some of the old products from that era will only have a master variant and not have a non-master variant, because it's before the point where having a non-master variant was made mandatory. And I'm not sure we even had proper soft-deletion on variants back then..?

Reply:
Yep, I just confirmed the hypothesis; lots of the earliest variants in the database on UK are master and have no corresponding  non-master siblings.

Reply:
Sooooo.... okay how about; we hunt down these ancient line items where the corresponding variant has `is_master: true` and we just flip the flag on it to `is_master: false` in those cases? The variants will be intact from the point of view of the line items and there's no loss of relevant information.